### PR TITLE
H_softMax for NLP

### PR DESCRIPTION
**Pull request by fouadb66**
I am trying to used your H_softmax implementation for multi-class text classification pb (# classes = 7000).
Currently I am using the soft max and categorical_crossentropy as a loss but the training is so slow.
my model takes as input the data of shape (X,30), 30 is the number of words in a sentence and X is the number of samples.  my label array is a one-hot encoded array of shape (X,7000).
I tried your implementation for my case but it dosent seems to work the loss still negative, I am assuming that I am using the hierarchical softmax layer correctly.
 Based on your dummy example (hsm_benchmark), my question is what would be the parameter of the fitting function in my case: (hsm_model.fit([noise_features, np.arange(n_examples)],
              np.zeros(noise_features.shape[0]), nb_epoch = 1)) .
Shall I add a soft max after this layer for the classification?  what about the loss ?

I will be waiting for your reply thank you so much

### CODE REVIEW

**Comment by fchollet** [comment]
Can someone please review this?

At first glance, I see that the API would need to be adapted for Keras 2.

**Comment by fchollet** [comment]
Some thoughts:

- it doesn't support TF
- the API would need to be updated
- an identity loss is not needed since we can have models that do not apply a loss to their outputs (as of Keras 2)

There are two ways we could introduce a potential HSoftmax layer:

- it could work by adding a loss via `self.add_loss` in `call`, require targets to be passed as a model `Input`, and require a `None` loss argument for the corresponding output
- we could try to make it work as a loss argument, by making it possible to have layer-like losses
